buffer/src/main/java/io/netty/buffer/UTF8Decoder.java

```diff
@@ -0,0 +1,210 @@
+        }
+        UTF8Processor processor = new UTF8Processor(new char[length]);
+
+        for (int i = 0; i < length; i++) {
+            processor.process(bytes[offset + i]);
+        }
+        return processor.toString();
+    }
+
+
+    public static String decode(DataInput input, int length) throws IOException {
+        UTF8Processor processor = new UTF8Processor(new char[length]);
+        for (int i = 0; i < length; i++) {
+            processor.process(input.readByte());
+        }
+        return processor.toString();
+    }
+
+    private static String illegalIndex(int offset, int length, int size) {
+        return "Illegal offset " + offset + " and length " + length + " for only " + size + " bytes!";
+    }
+
+    public static String decode(ByteBuf buf, int offset, int length) {
+        int size = buf.writerIndex();
+        if (MathUtil.isOutOfBounds(offset, length, size)) {
+            throw new IndexOutOfBoundsException(illegalIndex(offset, length, size));
+        }
+        UTF8Processor processor = new UTF8Processor(new char[length]);
+        buf.forEachByte(offset, length, processor);
+        return processor.toString();
+    }
+
+    public static final class UTF8Processor implements ByteProcessor {
+        private int state = UTF8_ACCEPT; // The state we are currently in, we start out by wanting a new character
+        private int codePoint; // Our current unicode code point
+        private int destLength;
+        private final char[] dest;
+
+        public UTF8Processor(char[] dest) {
+            this.dest = dest;
+        }
+
+        public boolean process(byte rawByte) {
+            int b = rawByte & 0xFF; // Byte.toUnsignedInt
+            if (b <= 127) {
+                processASCII(b);
+            } else {
+                processUnicode(b);
+            }
+            return true;
+        }
+
+        private void processASCII(int b) {
+            dest[destLength++] = (char) b;
+        }
+
+        private void processUnicode(int b) {
+            final int characterClass = CHARACTER_CLASSES[b];
+            if (state == UTF8_ACCEPT) {
+                codePoint = (0xFF >> characterClass) & b;
+            } else {
+                codePoint = (b & 0x3f) | (codePoint << 6);
+            }
+            switch (state = TRANSLATION_TABLE[state + characterClass]) {
+                case UTF8_REJECT:
+                    throw new IllegalArgumentException(illegalByte(b));
+                case UTF8_ACCEPT:
+                    // We're finished
+                    if (isBasicCodePoint(codePoint)) {
+                        // It can be represented as a single char
+                        dest[destLength++] = (char) codePoint;
+                    } else if (isValidCodePoint(codePoint)) {
+                        dest[destLength++] = highSurrogate(codePoint);
+                        dest[destLength++] = lowSurrogate(codePoint);
```

Comment:
Its impossible, since its  going to take at _least_ two bytes of input data to get a surrogate char. ASCII characters are never going to be surrogates :wink: